slowquant/qiskit_interface/circuit_wavefunction.py

```diff
@@ -1041,262 +947,173 @@ def print_progress_SPSA(
+        num_kappa = 0
+        if kappa_optimization:
+            num_kappa = len(self.kappa_idx)
+            self.kappa = parameters[:num_kappa]
+            self._move_cep()
+        if theta_optimization:
+            self.thetas = parameters[num_kappa:]
+        # Build operator
+        if theta_optimization:
+            H = hamiltonian_0i_0a(self.h_mo, self.g_mo, self.num_inactive_orbs, self.num_active_orbs)
+            H = H.get_folded_operator(self.num_inactive_orbs, self.num_active_orbs, self.num_virtual_orbs)
+            return self.QI.quantum_expectation_value(H)
+        rdms = ReducedDenstiyMatrix(
+            self.num_inactive_orbs,
+            self.num_active_orbs,
+            self.num_virtual_orbs,
+            rdm1=self.rdm1,
+            rdm2=self.rdm2,
+        )
+        return get_electronic_energy(rdms, self.h_mo, self.g_mo, self.num_inactive_orbs, self.num_active_orbs)
 
-    Return:
-        Electronic gradient with respect to orbital rotations.
-    """
-    rdms = ReducedDenstiyMatrix(
-        wf.num_inactive_orbs,
-        wf.num_active_orbs,
-        wf.num_active_orbs,
-        rdm1=wf.rdm1,
-        rdm2=wf.rdm2,
-    )
-    gradient = get_orbital_gradient(
-        rdms, wf.h_mo, wf.g_mo, wf.kappa_idx, wf.num_inactive_orbs, wf.num_active_orbs
-    )
-    return gradient
-
-
-def ansatz_parameters_gradient(
-    parameters: list[float], operator: FermionicOperator, quantum_interface: QuantumInterface
-) -> np.ndarray:
-    r"""Calculate gradient with respect to ansatz parameters.
+    def _calc_gradient_optimization(
+        self, parameters: list[float], theta_optimization: bool, kappa_optimization: bool
+    ) -> np.ndarray:
+        """Calculate electronic gradient.
 
-    Args:
-        parameters: Ansatz parameters.
-        operator: Operator which the derivative is with respect to.
-        quantum_interface: Interface to call quantum device.
+        Args:
+            parameters: Ansatz and orbital rotation parameters.
+            theta_optimization: Doing theta optimization.
+            kappa_optimization: Doing kappa optimization.
 
-    Returns:
-        Gradient with repsect to ansatz parameters.
-    """
-    gradient = np.zeros(len(parameters))
-    for i in range(len(parameters)):  # pylint: disable=consider-using-enumerate
-        R = quantum_interface.grad_param_R[quantum_interface.param_names[i]]
-        e_vals_grad = get_energy_evals_for_grad(operator, quantum_interface, parameters, i, R)
-        grad = 0.0
-        for j, mu in enumerate(list(range(1, 2 * R + 1))):
-            x_mu = (2 * mu - 1) / (2 * R) * np.pi
-            grad += e_vals_grad[j] * (-1) ** (mu - 1) / (4 * R * (np.sin(1 / 2 * x_mu)) ** 2)
-        gradient[i] = grad
-    return gradient
-
-
-def get_energy_evals_for_grad(
+        Returns:
+            Electronic gradient.
+        """
+        num_kappa = 0
+        gradient = np.zeros(len(parameters))
+        num_kappa = 0
+        if kappa_optimization:
```

Comment:
Because I do not assume any specific implementation details of the underlying optimizers.
In principle a gradient descent does not need to evaluate the energy but could just call the gradient many times.
In this case the current-expansion-point also need to be moved.

Reply:
Ok. And otherwise it would just mean that kappa = kappa_old and the expansion point is not actually moved when calling _move_cep?